eureka2-write-server/src/main/java/com/netflix/eureka2/server/channel/RegistrationChannelImpl.java

```diff
@@ -124,55 +125,81 @@ public Source getSource() {
      */
     @Override
     public Observable<Void> register(final InstanceInfo instanceInfo) {
-        if (!moveToState(STATE.Idle, STATE.Registered) &&
-                !moveToState(STATE.Registered, STATE.Registered)) {
-            STATE currentState = state.get();
-            if (STATE.Closed == currentState) {
-                // Since channel is already closed and hence the transport, we don't need to send an error on transport.
-                return Observable.error(CHANNEL_CLOSED_EXCEPTION);
-            } else {
-                Exception exception = new IllegalStateException("Unknown state error when registering, current state: " + currentState);
-                return sendErrorOnTransport(exception).doOnTerminate(new Action0() {
-                    @Override
-                    public void call() {
-                        close();
-                    }
-                });
-            }
+        if (moveToState(STATE.Idle, STATE.Registered)) {
+            return firstRegistration(instanceInfo);
```

Comment:
The returned result here is the chained result of registrationProcessor.register, whereas for update and unregister we have separate observables. What instead of setting up the initial registration processor in the first register, we do the set up independently and then can just use the same code as in registrationUpdate() for both the initial register and subsequent updates?